1-Draft/RFCXXXX-ScriptBlockLogging.md

```diff
@@ -0,0 +1,280 @@
+
+1. The transport to the Logging system is secure.
+   Secure includes:
+   - Encrypted
+   - Authenticated
+1. The machine admin is willing and authorized to send all ScriptBlocks to the logging system.
+
+### Extensions
+
+For this version of the feature,
+extension will be built into PowerShell.
+
+See [Extensions outside the engine](#extensions-outside-the-engine),
+for a discussion of future plans and
+concerns that need to be addressed.
+
+### Registration
+
+Only one logging extension can be registered at a time.
+There will be two options for registration,
+which are explained below:
+
+#### Registration cmdlet
+
+There will be a cmdlet to register a logging extension.
+`Register-PSScriptBlockLogger` will allow you to register a extension.
+It will update the Computer-Wide PowerShell configuration file to persist this as a policy.
+
+#### Policy
+
+There will be a new policy with three settings:
+
+- ScriptBlockLoggingExtension - The name of the extension, which should be a module installed on the system.
+- ScriptBlockLoggingExtensionKeyName - The name of [secret][secrets-rfc] storing the key used to access the extensions.
+- ScriptBlockLoggingExtensionProperties - Properties needed by the extension, persisted as JSON.
+
+### Extensions
+
+#### Loading the extension
+
+PowerShell will import the module and look at the Module Manifest PrivateData for what Class to use to send events.
+
+```PowerShell
+PrivateData = @{
+    PSData = @{
+        ScriptBlockLogging  = @{
+            TypeName='TypeName'
+        }
+    }
+}
+```
+
+The class must have a constructor with a parameter of type SecureString to pass the key and parameters for the rest of the Properties in `ScriptBlockLoggingExtensionProperties`.
+
+The class must have a void method named `PostScriptBlockLog` with the following parameters:
+
+PostLog(string scriptBlockText, string file, string scriptBlockHash, string parentScriptBlockHash, DateTime utcTime, string commandName, int runspaceId, string runspaceName)
+
+1. `scriptBlockText` of type string
+   - The text of the ScriptBlock being created
+1. `scriptBlockHash` of type string
+   - The hash (non-cryptographic) of the ScriptBlock being created
+1. `parentScriptBlockHash` of type string
+   - The hash (non-cryptographic) of the ScriptBlock which first used the created ScriptBlock
+1. `filePath` of type string
+   - If the ScriptBlock was from a file, the path to the file.
+1. `utcTime` of type DateTime
+   - UTC time the command was started.
+1. `commandName` of type string
+   - The first command name in the ScriptBlock
+1. `runspaceId` of type int
+   - a numeric identifier of the runspace the ScriptBlock was created in.
+1. `runspaceName` of type string
+   - The name the of the runspace the ScriptBlock was created in, if set.
+
+We will add code in `CompiledScriptBlock.LogScriptBlockCreation`,
+which does the work to load and call the extension.
+
+PowerShell will not retry calls to the extension which fail.
```

Comment:
We should log locally if an extension fails.  Do we no longer retry forever?

Reply:
```suggestion
PowerShell will not retry calls to the extension which fail and will log locally that the call failed.
```